### DISCUSSION TITLE
Koreader not indenting lists as expected in HTML5.css rendering only?

### DISCUSSION THREAD

**Discussion by Chairzard** [koreader/koreader#12909] [Q&A] [answered]
For context, I've been trying to improve the CSS of a book I pulled from Project Gutenberg using Calibre for use with KOReader.

It looks like KOReader is not indenting lists as expected while other epub viewers are (such as the viewer in Calibre), but only when using the html5.css style (however, the epub.css style works fine). Instead, lists remain unindented using the html5.css style. Is this a possible bug with the html5.css file KOReader uses, or am I missing something?

**Expected, indented rendering of indent for list in Calibre viewer:**

![Calibre](https://github.com/user-attachments/assets/addcf49e-1512-4835-a1f3-82eea7d2c3cb)

**Unexpected, unindented actual rendering of list in Koreader, using html5.css Style:**

![Koreader html5 css](https://github.com/user-attachments/assets/6f65284d-752f-445c-9dd0-91936c73dc37)

**(Working as expected) actual, indented rendered list in Koreader, using epub.css Style:**

![Koreader epub css](https://github.com/user-attachments/assets/0c5ba03e-00bf-4da6-855c-50dd82255695)

I double-checked using another book with (more complex) list tags and noticed the same issue happening:

**Expected, indented rendering of indent for list in Calibre viewer:**

![Book 2 Calibre](https://github.com/user-attachments/assets/366c9eeb-f59d-4290-aaac-e5f3fbd60b06)

**Unexpected, unindented actual rendering of list in Koreader, using html5.css Style:**

![Book 2 Koreader html5 css](https://github.com/user-attachments/assets/eff9fa0a-aaac-4fed-8416-69a2264b83b2)

**(Working as expected) actual, indented rendered list in Koreader, using epub.css Style:**

![Book 2 Koreader epub css](https://github.com/user-attachments/assets/bdebcdf4-af35-4099-89fd-c7acc63028a3)

I was poking around in the html5.css file (https://github.com/koreader/crengine/blob/master/cr3gui/data/html5.css)

**Comment by poire-z** [comment]
Not really sure I can notice what you expect.
But we do have a different behaviour than per-specs, we introduced and used a `padding-left: -cr-special`, which may acts on what you're observing.
All this is no longer clear in my head, so just giving you some pointers:
https://github.com/koreader/crengine/issues/521#issuecomment-1952341201
https://github.com/koreader/crengine/pull/555
https://github.com/koreader/crengine/blob/1aa6b9ecfd4d1ad0d7013ad72fab60f16691f7a7/cr3gui/data/html5.css#L484-L506
https://github.com/koreader/crengine/blob/1aa6b9ecfd4d1ad0d7013ad72fab60f16691f7a7/cr3gui/data/epub.css#L56-L61

The idea was to do better than the absurd default of `40px`, which should result in better layout - possibly not as how web browsers render, while still doing things ok if the publisher defines some stuff.

**Reply by Chairzard** [reply]
Thank you for the links!

I think the way that the epub.css is handling it in the [link you provided](https://github.com/koreader/crengine/blob/1aa6b9ecfd4d1ad0d7013ad72fab60f16691f7a7/cr3gui/data/epub.css#L56-L61) (adding `1em` to `margin-left` for `ol` and `ul`) is a perfect solution and that explains the difference I'm seeing as html5.css lacks that; do you think it would be possible to add that line to html5.css as well to unify the way that both styles handle this?

If the publisher doesn't define left margins for lists (this seems common), having `1em` in `margin-left` ensures that lists still look good and clearly shows that a section is a list section. Without that there, lists are not clearly defined with an indent in html5.css.

**Reply by Chairzard** [reply]
Here are some additional pics for clarity (see [my message above](https://github.com/koreader/koreader/discussions/12909#discussioncomment-11587060)):

_Without `1em` added:_

![Without 1em](https://github.com/user-attachments/assets/74ed496d-aa7a-433b-8e22-57a05e3926c3)

_With `1em` added (the red line shows the space added; this is how epub.css does it right now, but html5.css is currently missing this):_

![With 1em](https://github.com/user-attachments/assets/2b1bc14f-7167-4752-9a2d-d717832a1dfe)

**Reply by poire-z** [reply, marked as answer]
> (adding 1em to margin-left for ol and ul) is a perfect solution and that explains the difference I'm seeing as html5.css lacks that; do you think it would be possible to add that line to html5.css as well to unify the way that both styles handle this?

No, we should not :) or html5.css would no longer be "conforming to the current HTML standard/specs" - cf the big comment at its top.
https://html.spec.whatwg.org/multipage/rendering.html#lists does not specify margin-left or margin-inline-start.
So, we should not add it.
Its specifies `ol, ul { padding-inline-start: 40px; }`, which is a crap historical decision. It is what adds the spacing you want, but you'll get it only in some conditions - and possibly too pushed in or too pushed out in other conditions.
Our choice of `padding-inline-start: -cr-special` makes it work in all condition (never pushed in or out), but it doesn't give you the indentation you prefer. (Personally, I don't like that indentation - but if it's there and looks "ok" for you, it's luck with 40px; the right solution is to indeed use `margin-left: 1px`, but this is not in the HTML specs).
You can still help yourself with a style tweak to override our `-cr-special` with `40px` - or adds the `margin-left: 1em` you wish - but it should not be in html5.css.

**Reply by poire-z** [reply]
And to see why `padding-left: 40px` is crap, add it - and play with our `Zoom/DPI` bottom menu setting - that you would use to resize images, without afferct font size: the 40px will scale, you'll get a too large and ugly indentation.
`margin-left: 1em` would stay consistent with the font.
`padding-left: 1em` may also work, but if you have large item number, they could be larger than 1em and, with the default of `margin-left: 0`, they could be drawn outside this padding in the left margin, possibly outside the screen left edge and truncated. Our `padding-left: -cr-special` ensure this never happens.

**Reply by Chairzard** [reply]
Thank you again! The style tweak did the trick.
